src/mlpack/tests/main_tests/sparse_coding_test.cpp

```diff
@@ -462,4 +458,85 @@ BOOST_AUTO_TEST_CASE(SparseCodingModelReuseTest)
   CheckMatrices(codes, CLI::GetParam<arma::mat>("codes"));
 }
 
+/**
+ * Ensure that for different value of max iterations
+ * outputs are different.
+ */
+BOOST_AUTO_TEST_CASE(SparseCodingDiffMaxItrTest)
+{
+  mat inputData;
+  inputData.load("mnist_first250_training_4s_and_9s.arm");
+
+  // Shuffle input dataset.
+  inputData = shuffle(inputData);
+
+  // Generate test dataset.
+  mat testData;
+  testData = inputData.cols(450, 499);
+
+  // Generate train dataset.
+  inputData.shed_cols(450, 499);
+
+  // Generate initial dictionary.
+  SetInputParam("training", inputData);
+  SetInputParam("atoms", (int) 30);
+  SetInputParam("max_iterations", (int) 1);
+  SetInputParam("normalize", (bool) true);
+
+  mlpackMain();
+
+  mat initialDictionary =
+    std::move(CLI::GetParam<arma::mat>("dictionary"));
+
+  // Train for max_iterations equals to 2.
+
+  // Input data.
+  SetInputParam("training", inputData);
+  SetInputParam("atoms", (int) 30);
+  SetInputParam("initial_dictionary", initialDictionary);
+  SetInputParam("max_iterations", (int) 2);
+  SetInputParam("normalize", (bool) true);
+  SetInputParam("test", testData);
+
+  mlpackMain();
+
+  // Store outputs.
+  arma::mat dictionary =
+    std::move(CLI::GetParam<arma::mat>("dictionary"));
+  arma::mat codes =
+    std::move(CLI::GetParam<arma::mat>("codes"));
+
+  // Train for max_iterations equals to 100.
+
+  // Input data.
+  SetInputParam("training", std::move(inputData));
+  SetInputParam("atoms", (int) 30);
+  SetInputParam("initial_dictionary", std::move(initialDictionary));
+  SetInputParam("max_iterations", (int) 100);
+  SetInputParam("normalize", (bool) true);
+  SetInputParam("test", std::move(testData));
+
+  mlpackMain();
+
+  // Check that initial outputs and final outputs
+  // using two models model are different.
+  for (size_t i = 0; i < dictionary.n_elem; ++i)
+  {
+    if(dictionary[i]!=0 && CLI::GetParam<arma::mat>("dictionary")[i]!=0)
+    {
+      BOOST_REQUIRE_NE(dictionary[i],
+          CLI::GetParam<arma::mat>("dictionary")[i]);
```

Comment:
Currently I have used the above `if` condition to avoid failure when dictionary element is 0. Could you please suggest a better way so that it can be confirmed that atleast one element of the matrices is different. Using current code I am unable to do it properly as `BOOST_REQUIRE_NE` requires both element to be different.
One other way I tried to check that at least one element is different is through:
```
Log::Fatal.ignoreInput = true;
BOOST_REQUIRE_THROW(CheckMatrices(dictionary, CLI::GetParam<arma::mat>("dictionary")), std::runtime_error);
Log::Fatal.ignoreInput = false;
```
but it always reports the fatal error occured in `BOOST_REQUIRE_EQUAL`

Reply:
I think you could do

```
BOOST_REQUIRE_LT(arma::accu(dictionary == CLI::GetParam<arma::mat>("dictionary")), dictionary.n_elem);
```

In that the left hand side will be the number of elements that are exactly equal, and we want this to be less than the total number of elements in the matrix.  I think this will work fine; let me know if there is an issue.

Reply:
Wow! This is the finest way of doing that. Thanks @rcurtin.